File: trunk/EmberLib.net/EmberPlusRouter/Model/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace EmberPlusRouter.Model
{
   class Node : Element
   {
      public Node(int number, Element parent, string identifier)
      : base(number, parent, identifier)
      {
      }

      public override IEnumerable<Element> Children
      {
         get { return _children; }
      }

      public override int ChildrenCount
      {
         get { return _children.Count; }
      }

      public override void AddChild(Element child)
      {
         _children.Add(child);
      }

      public static Node CreateRoot()
      {
         return new Node(0, null, null);
      }

      public override TResult Accept<TState, TResult>(IElementVisitor<TState, TResult> visitor, TState state)
      {
         return visitor.Visit(this, state);
      }

      #region Implementation
      List<Element> _children = new List<Element>();
      #endregion
   }
}

File: branches/embersdk-1-2-branch/tinyember/TinyEmberPlusRouter/TinyEmberPlusRouter/model/Node.cpp
#include "Node.h"
#include "ElementVisitor.h"

namespace model
{
   Node::Node(int number, Element* parent, std::string const& identifier)
      : Element(number, parent, identifier)
   {}

   // override
   void Node::accept(ElementVisitor* visitor)
   {
      visitor->visit(this);
   }
}
